File: src/tpuff/commands/schema.py
"""Schema management commands for tpuff CLI."""

import json
import sys
from dataclasses import dataclass, field

import click
from rich.console import Console

from tpuff.client import get_namespace

console = Console()


@dataclass
class SchemaDiff:
    """Result of comparing two schemas."""

    unchanged: dict[str, str] = field(default_factory=dict)
    additions: dict[str, str] = field(default_factory=dict)
    conflicts: dict[str, tuple[str, str]] = field(default_factory=dict)  # attr -> (old_type, new_type)

    @property
    def has_conflicts(self) -> bool:
        """Check if there are any type conflicts."""
        return len(self.conflicts) > 0

    @property
    def has_changes(self) -> bool:
        """Check if there are any additions or conflicts."""
        return len(self.additions) > 0 or len(self.conflicts) > 0


def normalize_schema_type(attr_type: object) -> str:
    """Normalize a schema type to a comparable string representation.

    Handles both simple string types and complex type objects from turbopuffer.
    """
    if hasattr(attr_type, "model_dump"):
        # Pydantic model - convert to JSON string for comparison
        return json.dumps(attr_type.model_dump(), sort_keys=True)
    elif hasattr(attr_type, "to_dict"):
        return json.dumps(attr_type.to_dict(), sort_keys=True)
    elif isinstance(attr_type, dict):
        return json.dumps(attr_type, sort_keys=True)
    else:
        return str(attr_type)


def schema_type_for_display(attr_type: object) -> str:
    """Convert a schema type to a human-readable display string."""
    if hasattr(attr_type, "model_dump"):
        dumped = attr_type.model_dump()
        # For complex types, show the full dict; for simple, just the string
        if isinstance(dumped, dict) and len(dumped) == 1 and "type" in dumped:
            return str(dumped["type"])
        return json.dumps(dumped)
    elif hasattr(attr_type, "to_dict"):
        return json.dumps(attr_type.to_dict())
    elif isinstance(attr_type, dict):
        if len(attr_type) == 1 and "type" in attr_type:
            return str(attr_type["type"])
        return json.dumps(attr_type)
    else:
        return str(attr_type)


def display_schema_diff(diff: SchemaDiff, namespace: str) -> None:
    """Display a schema diff with Rich formatting.

    Args:
        diff: The computed schema diff
        namespace: The namespace name (for header)
    """
    console.print(f"\n[bold]Schema changes for namespace: {namespace}[/bold]\n")

    if not diff.has_changes and not diff.unchanged:
        console.print("[dim]No schema attributes[/dim]")
        return

    # Sort all attributes for consistent output
    all_attrs = sorted(
        set(diff.unchanged.keys()) | set(diff.additions.keys()) | set(diff.conflicts.keys())
    )

    for attr in all_attrs:
        if attr in diff.unchanged:
            # Unchanged attribute
            console.print(f"  {attr}: {diff.unchanged[attr]}")
        elif attr in diff.additions:
            # New attribute
            console.print(f"[green]+{attr}: {diff.additions[attr]}[/green]  [dim](new)[/dim]")
        elif attr in diff.conflicts:
            # Type conflict
            old_type, new_type = diff.conflicts[attr]
            console.print(
                f"[red]!{attr}: {old_type} -> {new_type}[/red]  "
                f"[dim](type change not allowed)[/dim]"
            )

    console.print()  # Blank line at end


def compute_schema_diff(
    current_schema: dict[str, object] | None,
    new_schema: dict[str, object],
) -> SchemaDiff:
    """Compute the difference between current and new schemas.

    Args:
        current_schema: The existing schema (None if namespace doesn't exist)
        new_schema: The schema to be applied

    Returns:
        SchemaDiff with unchanged, additions, and conflicts
    """
    diff = SchemaDiff()

    if current_schema is None:
        current_schema = {}

    # Normalize current schema for comparison
    current_normalized = {
        attr: normalize_schema_type(attr_type)
        for attr, attr_type in current_schema.items()
    }

    # Compare each attribute in the new schema
    for attr, new_type in new_schema.items():
        new_type_normalized = normalize_schema_type(new_type)
        new_type_display = schema_type_for_display(new_type)

        if attr not in current_normalized:
            # New attribute
            diff.additions[attr] = new_type_display
        elif current_normalized[attr] == new_type_normalized:
            # Unchanged
            diff.unchanged[attr] = new_type_display
        else:
            # Type conflict
            old_type_display = schema_type_for_display(current_schema[attr])
            diff.conflicts[attr] = (old_type_display, new_type_display)

    return diff


@click.group("schema", context_settings={"help_option_names": ["-h", "--help"]})
def schema() -> None:
    """Manage namespace schemas."""
    pass


@schema.command("get", context_settings={"help_option_names": ["-h", "--help"]})
@click.option("-n", "--namespace", required=True, help="Namespace to get schema from")
@click.option("-r", "--region", help="Override the region (e.g., aws-us-east-1, gcp-us-central1)")
@click.option("--raw", is_flag=True, help="Output raw JSON without formatting (for piping)")
@click.pass_context
def schema_get(
    ctx: click.Context,
    namespace: str,
    region: str | None,
    raw: bool,
) -> None:
    """Display the schema for a namespace."""
    try:
        ns = get_namespace(namespace, region)
        metadata = ns.metadata()

        # Extract schema from metadata
        schema_data = metadata.schema if hasattr(metadata, "schema") else {}

        if not schema_data:
            if raw:
                print("{}")
            else:
                console.print(f"[yellow]No schema found for namespace: {namespace}[/yellow]")
            return

        # Convert schema to serializable format
        schema_dict = {}
        for attr_name, attr_type in schema_data.items():
            # Handle both simple string types and complex type objects
            if hasattr(attr_type, "model_dump"):
                schema_dict[attr_name] = attr_type.model_dump()
            elif hasattr(attr_type, "to_dict"):
                schema_dict[attr_name] = attr_type.to_dict()
            else:
                schema_dict[attr_name] = str(attr_type)

        if raw:
            print(json.dumps(schema_dict))
        else:
            console.print(f"\n[bold]Schema for namespace: {namespace}[/bold]\n")
            console.print(json.dumps(schema_dict, indent=2))

    except Exception as e:
        if raw:
            print(json.dumps({"error": str(e)}), file=sys.stderr)
        else:
            console.print(f"[red]Error: {e}[/red]")
        sys.exit(1)

File: internal/schema/schema.go
package schema

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
)

// Valid simple schema types.
var ValidSimpleTypes = map[string]bool{
	"string":   true,
	"[]string": true,
	"uint64":   true,
	"int":      true,
	"float":    true,
	"uuid":     true,
	"bool":     true,
}

// VectorTypePattern matches vector types like [1536]f32 or [768]f16.
var VectorTypePattern = regexp.MustCompile(`^\[\d+\]f(16|32)$`)

// Valid keys for complex type objects.
var ValidTypeKeys = map[string]bool{
	"type":             true,
	"full_text_search": true,
	"regex_index":      true,
	"filterable":       true,
	"ann":              true,
	"glob":             true,
	"regex":            true,
}

// PropertyChange represents a single property change within an attribute.
type PropertyChange struct {
	Key      string
	OldValue string
	NewValue string
}

// AttrModification holds the details of a modified attribute.
type AttrModification struct {
	BaseType string
	Changes  []PropertyChange
}

// SchemaDiff holds the result of comparing two schemas.
type SchemaDiff struct {
	Unchanged     map[string]string         // attr -> display type
	Additions     map[string]string         // attr -> display type (new fields)
	Modifications map[string]AttrModification // attr -> modification details
	Conflicts     map[string][2]string      // attr -> [old_type, new_type] (base type change — blocked)
}

// HasConflicts returns true if there are any type conflicts.
func (d *SchemaDiff) HasConflicts() bool {
	return len(d.Conflicts) > 0
}

// HasModifications returns true if there are any property modifications.
func (d *SchemaDiff) HasModifications() bool {
	return len(d.Modifications) > 0
}

// HasChanges returns true if there are any additions, modifications, or conflicts.
func (d *SchemaDiff) HasChanges() bool {
	return len(d.Additions) > 0 || len(d.Modifications) > 0 || len(d.Conflicts) > 0
}

// NormalizeSchemaType normalizes a schema type to a comparable string.
// Handles strings, map[string]any, and other types.
func NormalizeSchemaType(attrType any) string {
	switch v := attrType.(type) {
	case string:
		return v
	case map[string]any:
		b, _ := json.Marshal(v)
		return string(b)
	default:
		return fmt.Sprintf("%v", v)
	}
}

// SchemaTypeForDisplay converts a schema type to a human-readable string.
func SchemaTypeForDisplay(attrType any) string {
	switch v := attrType.(type) {
	case string:
		return v
	case map[string]any:
		if len(v) == 1 {
			if t, ok := v["type"]; ok {
				return fmt.Sprintf("%v", t)
			}
		}
		b, _ := json.Marshal(v)
		return string(b)
	default:
		return fmt.Sprintf("%v", v)
	}
}

// ValidateSchemaType validates a single schema attribute type.
// Returns a list of error messages (empty if valid).
func ValidateSchemaType(attrName string, attrType any) []string {
	var errors []string

	switch v := attrType.(type) {
	case string:
		if ValidSimpleTypes[v] {
			return nil
		}
		if VectorTypePattern.MatchString(v) {
			return nil
		}
		validTypes := sortedSimpleTypes()
		errors = append(errors, fmt.Sprintf(
			"Attribute '%s': invalid type '%s'. Valid types: %s, or vector format [dims]f32/f16",
			attrName, v, strings.Join(validTypes, ", ")))

	case map[string]any:
		typeVal, hasType := v["type"]
		if !hasType {
			errors = append(errors, fmt.Sprintf(
				"Attribute '%s': complex type object must have a 'type' key", attrName))
		} else {
			baseType, isStr := typeVal.(string)
			if !isStr {
				errors = append(errors, fmt.Sprintf(
					"Attribute '%s': 'type' must be a string", attrName))
			} else if !ValidSimpleTypes[baseType] && !VectorTypePattern.MatchString(baseType) {
				validTypes := sortedSimpleTypes()
				errors = append(errors, fmt.Sprintf(
					"Attribute '%s': invalid base type '%s'. Valid types: %s, or vector format [dims]f32/f16",
					attrName, baseType, strings.Join(validTypes, ", ")))
			}
		}

		// Check for unknown keys
		var unknownKeys []string
		for k := range v {
			if !ValidTypeKeys[k] {
				unknownKeys = append(unknownKeys, k)
			}
		}
		if len(unknownKeys) > 0 {
			sort.Strings(unknownKeys)
			validKeys := sortedValidTypeKeys()
			errors = append(errors, fmt.Sprintf(
				"Attribute '%s': unknown keys %v. Valid keys: %s",
				attrName, unknownKeys, strings.Join(validKeys, ", ")))
		}

		// Validate specific option types
		if fts, ok := v["full_text_search"]; ok {
			switch fts.(type) {
			case bool:
				// ok
			case map[string]any:
				// ok — full config object from API
			default:
				errors = append(errors, fmt.Sprintf(
					"Attribute '%s': 'full_text_search' must be a boolean or object", attrName))
			}
		}
		if ri, ok := v["regex_index"]; ok {
			if _, isBool := ri.(bool); !isBool {
				errors = append(errors, fmt.Sprintf(
					"Attribute '%s': 'regex_index' must be a boolean", attrName))
			}
		}
		if f, ok := v["filterable"]; ok {
			if _, isBool := f.(bool); !isBool {
				errors = append(errors, fmt.Sprintf(
					"Attribute '%s': 'filterable' must be a boolean", attrName))
			}
		}

	default:
		errors = append(errors, fmt.Sprintf(
			"Attribute '%s': type must be a string or object, got %T", attrName, attrType))
	}

	return errors
}

// ValidateSchema validates a complete schema dictionary.
// Returns a list of error messages (empty if valid).
func ValidateSchema(schemaData map[string]any) []string {
	var errors []string

	for attrName, attrType := range schemaData {
		if attrName == "" {
			errors = append(errors, "Attribute name cannot be empty")
			continue
		}
		errors = append(errors, ValidateSchemaType(attrName, attrType)...)
	}

	// The "id" attribute must have filterable: true if present.
	if idType, ok := schemaData["id"]; ok {
		if m, isMap := idType.(map[string]any); isMap {
			if f, hasFilt := m["filterable"]; hasFilt {
				if b, isBool := f.(bool); isBool && !b {
					errors = append(errors, "Attribute 'id': must have \"filterable\" set to true")
				}
			}
		}
	}

	// Vector attributes must have ann: true.
	for attrName, attrType := range schemaData {
		m, isMap := attrType.(map[string]any)
		if !isMap {
			continue
		}
		baseType, _ := m["type"].(string)
		if !VectorTypePattern.MatchString(baseType) {
			continue
		}
		ann, hasAnn := m["ann"]
		if !hasAnn {
			errors = append(errors, fmt.Sprintf(
				"Attribute '%s': vector type must have \"ann\" set to true", attrName))
		} else if b, isBool := ann.(bool); isBool && !b {
			errors = append(errors, fmt.Sprintf(
				"Attribute '%s': vector type must have \"ann\" set to true", attrName))
		}
	}

	return errors
}

// ExtractBaseType returns the base type string from a schema attribute value.
// For simple strings like "string", returns the string itself.
// For complex objects like {"type":"string","filterable":true}, returns "string".
func ExtractBaseType(attrType any) string {
	switch v := attrType.(type) {
	case string:
		return v
	case map[string]any:
		if t, ok := v["type"].(string); ok {
			return t
		}
		return ""
	default:
		return fmt.Sprintf("%v", v)
	}
}

// ParseVectorDims extracts the dimension count from a vector type string like "[768]f32".
func ParseVectorDims(vectorType string) int {
	matches := VectorTypePattern.FindStringSubmatch(vectorType)
	if matches == nil {
		return 0
	}
	// Extract number between brackets
	start := 1 // skip '['
	end := strings.Index(vectorType, "]")
	if end <= start {
		return 0
	}
	var dims int
	fmt.Sscanf(vectorType[start:end], "%d", &dims)
	return dims
}

// annEnabled returns true if the ann value represents "enabled" (either bool true or a config object).
func annEnabled(v any) bool {
	switch a := v.(type) {
	case bool:
		return a
	case map[string]any:
		// Any non-empty object means ann is enabled
		return len(a) > 0
	}
	return false
}

// computePropertyChanges compares two attribute maps and returns the changed properties.
// attrName is used for special-case handling (e.g. "id").
func computePropertyChanges(attrName string, oldMap, newMap map[string]any) []PropertyChange {
	var changes []PropertyChange
	// Collect all keys
	allKeys := make(map[string]bool)
	for k := range oldMap {
		allKeys[k] = true
	}
	for k := range newMap {
		allKeys[k] = true
	}

	sorted := make([]string, 0, len(allKeys))
	for k := range allKeys {
		sorted = append(sorted, k)
	}
	sort.Strings(sorted)

	for _, k := range sorted {
		if k == "type" {
			continue // base type is shown separately
		}

		// The "id" attribute is always filterable regardless of what the API reports.
		if attrName == "id" && k == "filterable" {
			continue
		}

		// For "ann", compare by enabled/disabled rather than exact representation,
		// since the API returns an object but writes accept a boolean.
		if k == "ann" {
			if annEnabled(oldMap[k]) != annEnabled(newMap[k]) {
				oldVal, _ := json.Marshal(oldMap[k])
				newVal, _ := json.Marshal(newMap[k])
				changes = append(changes, PropertyChange{
					Key:      k,
					OldValue: string(oldVal),
					NewValue: string(newVal),
				})
			}
			continue
		}

		oldVal, _ := json.Marshal(oldMap[k])
		newVal, _ := json.Marshal(newMap[k])
		if string(oldVal) != string(newVal) {
			changes = append(changes, PropertyChange{
				Key:      k,
				OldValue: string(oldVal),
				NewValue: string(newVal),
			})
		}
	}
	return changes
}

// toMap normalizes an attribute type to a map for property comparison.
func toMap(attrType any) map[string]any {
	switch v := attrType.(type) {
	case map[string]any:
		return v
	case string:
		return map[string]any{"type": v}
	default:
		return map[string]any{"type": fmt.Sprintf("%v", v)}
	}
}

// ComputeSchemaDiff computes the difference between current and new schemas.
// currentSchema can be nil (namespace doesn't exist yet).
func ComputeSchemaDiff(currentSchema, newSchema map[string]any) SchemaDiff {
	diff := SchemaDiff{
		Unchanged:     make(map[string]string),
		Additions:     make(map[string]string),
		Modifications: make(map[string]AttrModification),
		Conflicts:     make(map[string][2]string),
	}

	if currentSchema == nil {
		currentSchema = map[string]any{}
	}

	// Normalize current schema
	currentNormalized := make(map[string]string)
	for attr, attrType := range currentSchema {
		currentNormalized[attr] = NormalizeSchemaType(attrType)
	}

	// Compare each attribute in new schema
	for attr, newType := range newSchema {
		newTypeNormalized := NormalizeSchemaType(newType)
		newTypeDisplay := SchemaTypeForDisplay(newType)

		oldNorm, exists := currentNormalized[attr]
		if !exists {
			diff.Additions[attr] = newTypeDisplay
		} else if oldNorm == newTypeNormalized {
			diff.Unchanged[attr] = newTypeDisplay
		} else {
			oldTypeDisplay := SchemaTypeForDisplay(currentSchema[attr])
			oldBase := ExtractBaseType(currentSchema[attr])
			newBase := ExtractBaseType(newType)
			if oldBase == newBase || oldBase == "unknown" || oldBase == "[]unknown" {
				// Same base type, different properties — check for meaningful changes
				changes := computePropertyChanges(attr, toMap(currentSchema[attr]), toMap(newType))
				if len(changes) > 0 {
					diff.Modifications[attr] = AttrModification{
						BaseType: newBase,
						Changes:  changes,
					}
				} else {
					// All differences were suppressed (e.g. ann normalization)
					diff.Unchanged[attr] = newTypeDisplay
				}
			} else {
				// Different base type — conflict (blocked)
				diff.Conflicts[attr] = [2]string{oldTypeDisplay, newTypeDisplay}
			}
		}
	}

	return diff
}

// LoadSchemaFile loads and validates a schema from a JSON file.
func LoadSchemaFile(filePath string) (map[string]any, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("Schema file not found: %s", filePath)
		}
		return nil, fmt.Errorf("Error reading schema file: %w", err)
	}

	var raw any
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("Invalid JSON in schema file: %s", err)
	}

	schemaData, ok := raw.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("Schema file must contain a JSON object")
	}

	errors := ValidateSchema(schemaData)
	if len(errors) > 0 {
		return nil, fmt.Errorf("Invalid schema:\n  %s", strings.Join(errors, "\n  "))
	}

	return schemaData, nil
}

func sortedSimpleTypes() []string {
	types := make([]string, 0, len(ValidSimpleTypes))
	for t := range ValidSimpleTypes {
		types = append(types, t)
	}
	sort.Strings(types)
	return types
}

func sortedValidTypeKeys() []string {
	keys := make([]string, 0, len(ValidTypeKeys))
	for k := range ValidTypeKeys {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}
